File: Homework/Van_Le/lesson7/HouseRobberIII/HouseRobberIII.cs
namespace LeetCodeSolutions
{
    public class HouseRobberIII
    {
        Dictionary<TreeNode, int> robResult = new Dictionary<TreeNode, int>();
        Dictionary<TreeNode, int> notRobResult = new Dictionary<TreeNode, int>();

        public int Helper(TreeNode node, bool parentRobbed)
        {
            if (node == null)
            {
                return 0;
            }
            if (parentRobbed)
            {
                if (robResult.ContainsKey(node))
                {
                    return robResult[node];
                }
                int result = Helper(node.left, false) + Helper(node.right, false);
                robResult[node] = result;
                return result;
            }
            else
            {
                if (notRobResult.ContainsKey(node))
                {
                    return notRobResult[node];
                }
                int rob = node.val + Helper(node.left, true) + Helper(node.right, true);
                int notRob = Helper(node.left, false) + Helper(node.right, false);
                int result = Math.Max(rob, notRob);
                notRobResult[node] = result;
                return result;
            }
        }

        public int Rob(TreeNode root)
        {
            return Helper(root, false);
        }
    }
}

File: Homework/Quang_Minh/HomeWorkLesson7/HouseRobberIII.java
package HomeWorkLesson7;

public class HouseRobberIII {
    public int rob(TreeNode root) {
        if (root == null) return 0;
        return dfs(root, true);
    }

    public int dfs(TreeNode root, boolean canRob) {
        if (root == null) return 0;
        int rob = root.val + dfs(root.left, false) + dfs(root.right, false);
        int noRob = dfs(root.left, true) + dfs(root.right, true);
        if (canRob) {
            return Math.max(rob, noRob);
        } else {
            return noRob;
        }
    }
}
